docs/en/contributor/Code-Conventions.md

```diff
@@ -123,39 +94,436 @@ public MyClass {
 if (LOG.isDebugEnabled()) {
-    LOG.debug("Failed to connect to address {} due to exception", host + ":" + port, e); // OK
+    LOG.debug("Failed to connect to address {} due to exception", host + ":" + port, e);
 }
+
+// Not recommended: string concatenation is always performed
+LOG.debug("Failed to connect to {} due to exception", host + ":" + port, e);
+```
+
+**When to Use**
+* Any exit or entry point of the process where execution is handed to an external process
+  * Example: A request is made to the under file system
+
+**During Exception Handling**
+* Debug level logs provide the control flow logging, in the case of an error exit, the exception
+message is logged
+* For warn level messages which may not always be benign, the stack trace can be logged at the
+debug level. These debug level stack traces should eventually be phased out.
+
+#### Trace Log Level
+
+**When to Use**
+* Any degree of detail within an experimental or new feature
+
+**During Exception Handling**
+* Up to the developer’s discretion
+
+## Exceptions
+
+### Throwing Exceptions
+
+#### Use unchecked exceptions for events which indicate bugs
+
+Examples: Illegal states, invalid API usage
+
+```java
+public void method(String neverSupposedToBeNull) {
+  Preconditions.checkNotNull(neverSupposedToBeNull, "neverSupposedToBeNull");
+}
+```
+
+#### Use checked exceptions for everything else
+
+Examples: File not found, interrupted exception, timeout exceeded
+
+```java
+// Recommended: the checked exception should just be propagated
+public void handleRawUserInput(String date) throws InvalidDateException {
+  parseDate(date);
+}
+
+// Don't do this - it's reasonable for user input to be invalid.
+public void handleRawUserInput(String date) throws InvalidDateException {
+  try {
+    parseDate(date);
+  } catch (InvalidDateException e) {
+    throw new RuntimeExcepiton("date " + date + " is invalid", e);
+  }
+}
+```  
+
+#### Design code to minimize use of checked exceptions
+
+Require callers to validate inputs so that invalid arguments can be unchecked exceptions.
+
+#### Reuse existing exceptions when appropriate
+
+Try to find an appropriate existing exception before inventing a new one.
+
+#### Selectively javadoc exceptions
+Only write exception javadoc when you think it will be useful to the developer using the method.
+There are so many sources of IOException that it's almost never useful to include javadoc for it.
+
+#### AlluxioStatusExceptions
+
+On the wire we represent exceptions with one of 14 status codes, e.g. NOT_FOUND, UNAVAILABLE. Within
+our server and client code, we represent these exceptions using exception classes corresponding to
+these statuses, e.g. NotFoundException and UnavailableException. AlluxioStatusException is the
+superclass for these Java exceptions.
+
+### Handling exceptions
+
+#### Never swallow an unchecked exception.
+
+Either log it, propagate it, or log it and kill the system
```

Comment:
I wrote this originally but looking back I think it's better to just say "Either log the exception or propagate it"

Reply:
Done